### DISCUSSION TITLE
.NET 4 and NativeAOT

### DISCUSSION THREAD

**Discussion by klimatr26** [itsmattkc/dotnet9x#18] [Q&A]
Great work!

I think it would be harder to do the same with .NET Framework 4. But I wanted to ask: Do you think it would be possible to backport newer versions of CLR 4.0 (like .NET Framework 4.8) to at least Windows XP, since 4.0 already supports XP SP3?

I like things like async/await, Span and SSE support from newer versions of the .NET Framework and .NET.

Also, maybe NativeAOT of .NET 8 or 9 (where I think they are working on implementing Windows x86 support) could be made to work with Windows XP or at least Windows Vista? I say that because .NET 8 only officially supports Windows 10+, but programs compiled with NativeAOT seem to work on Windows 7 SP1 (without installing anything if you statically link UCRT).

Also, I'm guessing WPF doesn't work on Windows 9x?

Again, all that you've done is amazing, and the only other options I've seen for 9x support are the Rust9x project and some post about using Clang to compile for NT 4.0 that doesn't seem to work well under Windows 9x (or you could use an older compiler and an older version of C++, and something like OpenWatcom, maybe, but C# and modern C++ feel easier to use for someone like me).

**Comment by Fancy2209** [comment]
I think .NET 4 would be easier by doing a port of [Mono](https://github.com/mono/mono)
Mono9x would be amazing though, Imagine FEZ or Celeste on Windows 9x!

**Reply by itsmattkc** [reply]
Yeah this is what I'm thinking too, .NET 2.0-3.5 were only "easy" because the gap was (technically) only 95 ➜ 98. If we're making bigger jumps, I think we'll be wanting to make a source port instead of something like Mono or .NET Core.

**Reply by Fancy2209** [reply]
Since Mono was actually officially built to 9x and 2000 before (hasn't been for many years though), it'd make more sense I think
From their technical FAQ
https://www.mono-project.com/docs/faq/technical/#can-mono-run-on-windows-9x-or-me-editions

**Comment by MaxNiftyNine** [comment]
It would be sick if we could get unity games running because they use .Net 4 I think but theres probably more to that

**Reply by klimatr26** [reply]
I think old Unity games used Mono, while new games are usually compiled ahead of time using IL2CPP or similar. I don't think they have ever used .NET Framework. Also, other parts of the engine are written in -I think- C++. Maybe porting something like Mono could make games like Sonic Runners work there, except for the fact that many DirectX and OpenGL functions are probably missing too (especially on Windows 95) (also, you would probably need newer graphics drivers). But maybe XNA games could be made to work?
